src/amo/components/SearchForm/index.js

```diff
@@ -0,0 +1,215 @@
+  ADDON_TYPE_EXTENSION,
+  ADDON_TYPE_THEME,
+  validAddonTypes,
+} from 'core/constants';
+import translate from 'core/i18n/translate';
+import { convertFiltersToQueryParams } from 'core/searchUtils';
+import SearchInput from 'ui/components/SearchInput';
+import LoadingText from 'ui/components/LoadingText';
+import { withErrorHandler } from 'core/errorHandler';
+import {
+  autocompleteCancel,
+  autocompleteStart,
+} from 'core/reducers/autocomplete';
+import Suggestion from 'amo/components/SearchForm/Suggestion';
+
+import 'core/css/inc/lib.scss';
+import './styles.scss';
+
+
+export class SearchFormBase extends React.Component {
+  static propTypes = {
+    addonType: PropTypes.string,
+    api: PropTypes.object.isRequired,
+    i18n: PropTypes.object.isRequired,
+    pathname: PropTypes.string.isRequired,
+    query: PropTypes.string.isRequired,
+    dispatch: PropTypes.func.isRequired,
+    errorHandler: PropTypes.object.isRequired,
+    suggestions: PropTypes.arrayOf(PropTypes.shape({
+      name: PropTypes.string.isRequired,
+      url: PropTypes.string.isRequired,
+      iconUrl: PropTypes.string.isRequired,
+    })).isRequired,
+    loadingSuggestions: PropTypes.bool.isRequired,
+  }
+
+  static contextTypes = {
+    router: PropTypes.object,
+  }
+
+  constructor(props: Object) {
+    super(props);
+
+    this.state = {
+      searchValue: props.query || '',
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { query } = nextProps;
+
+    if (this.props.query !== query) {
+      this.setState({ searchValue: query || '' });
+    }
+  }
+
+  getSuggestions() {
+    if (this.props.loadingSuggestions) {
+      return Array(10).fill('loading-text-will-be-rendered');
+    }
+
+    return this.props.suggestions;
+  }
+
+  goToSearch(query) {
+    const { addonType, api, pathname } = this.props;
+    const filters = { query };
+
+    if (addonType) {
+      filters.addonType = addonType;
+    }
+
+    this.context.router.push({
+      pathname: `/${api.lang}/${api.clientApp}${pathname}`,
+      query: convertFiltersToQueryParams(filters),
+    });
+  }
+
+  handleSearch = (e) => {
+    e.preventDefault();
+    this.searchInput.blur();
+    this.goToSearch(this.searchInput.value);
+  }
+
+  handleSearchChange = (e) => {
+    this.setState({ searchValue: e.target.value });
+  }
+
+  handleSuggestionsFetchRequested = ({ value }) => {
+    if (!value) {
+      return;
+    }
+
+    const { addonType, dispatch, errorHandler } = this.props;
+    const filters = { query: value };
+
+    if (addonType) {
+      filters.addonType = addonType;
+    }
+
+    dispatch(autocompleteStart({
+      errorHandlerId: errorHandler.id,
+      filters,
+    }));
+  }
+
+  handleSuggestionsClearRequested = () => {
+    this.props.dispatch(autocompleteCancel());
+  }
+
+  handleSuggestionSelected = (e, { suggestion }) => {
+    e.preventDefault();
+    this.setState({ searchValue: '' }, () => {
+      this.context.router.push(suggestion.url);
```

Comment:
That seems like something we should at least have a test for (is there one?)... if that behaviour changed it would cause some weird bugs.